### DISCUSSION TITLE
Make `MemberNotNullAttribute` and `MemberNotNullWhenAttribute` work with async methods

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/csharplang#4994] [Language Ideas]
### Description

Proposal is to enforce `Task` returned from `async` method marked with `MemberNotNull` to be completed before compiler will consider members to be initialized.

Here's the code that illustrates the problem:

```cs
class C
{
    string? field;

    async Task M()
    {
        // nullability analysis currently doesn't enforce InitAsync's Task to be completed in some way
        // before it considers field to be initialized
        await InitAsync(); 
    }

    [MemberNotNull(nameof(field))]
    async Task InitAsync()
    {
        // If InitAsync is not awaited, field won't be initialized.
#pragma warning disable CS8774 // Member must have a non-null value when exiting.
        field ??= await GetValueAsync();
#pragma warning restore CS8774 // Member must have a non-null value when exiting.
    }
}

```

### Regression?

No

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-threading-tasks
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

### Description

Proposal is to enforce `Task` returned from `async` method marked with `MemberNotNull` to be completed before compiler will consider members to be initialized.

Here's the code that illustrates the problem:

```cs
class C
{
    string field;
    async Task M()
    {
        // nullability analysis currently doesn't enforce InitAsync's Task to be completed in some way
        // before it considers field to be initialized
        await InitAsync(); 
    }

    [MemberNotNull(nameof(field))]
    async Task InitAsync()
    {
        // If InitAsync is not awaited, field won't be initialized.
#pragma warning disable CS8774 // Member must have a non-null value when exiting.
        field ??= await GetValueAsync();
#pragma warning restore CS8774 // Member must have a non-null value when exiting.
    }
}

```

### Regression?

No


<table>
  <tr>
    <th align="left">Author:</th>
    <td>maxkoshevoi</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Threading.Tasks`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by Joe4evr** [comment]
This happens because awaiting is technically optional, and when you hit an `await` in your *called* method it'll return back to you while the operation is still pending. Consider:
```cs
async Task M()
{
    Task t = InitAsync();
    string trim = filed.Trim(); //oops, field was still null because the task was not yet completed
    await t;
}
```
Solving this would at least require new attributes to indicate `MemberNotNullAfterAwait` for async methods, but that'll only work for the case where the method is immediately awaited (admittedly, that's the most common thing). The other case would require more complex flow tracking, but that's extremely difficult, and even if it was possible, it'd slow down the compiler to a crawl.

**Comment by maxkoshevoi** [comment]
@Joe4evr If complete tracking of returned task would be too much for a compiler, could it at least check that task is being completed in the same statement where method is called (which will be most common case):
- `await InitAsync()` 
- `InitAsync().Result`

Does this really require a new attribute? Compiler can just look at the return type of the method and switch behavior based on it.
I don't think there are methods that are initializing members synchronously and return `Task`. I don't oppose new attribute, just think it will be simpler with a single one.

**Comment by Kuinox** [comment]
I proposed a new Attribute for this here: https://github.com/dotnet/csharplang/discussions/5657
